README.md
# chase
A simple game where you need to escape from an increasing number of enemys


CHANGELOG.md
#### 17/05/2022 - Gustavo
## Added
   - Ahora los botones de las bombas y teleports reflejan correctamente el stock de cada elemento
   - El boton de guided-teleport hace que la flecha del mouse se transforme en una mira

### Notes
    - Intenté, sin exito, hacer una clase que cambie la pinta del boton guded-teleport para que se sepa que ese modo está seleccionado. 
      Hice una clase llamada 'button-selected' que tendrías que revisar/corregir
    - Hice tambien una clase 'button-hidden' que funciona pero igual habría que revisarla por si no cumple con las buenas prácticas
    - Hacer clases/elementos para que TODOS los botones de herramientas puedan tener entre 0 y 5 unidades, o bien infinitos, o bien no 
      existir (button-hidden) 
    - Ahora todas las clases CSS y elementos HTML pueden tener la convension de nombres que se estila para cada caso, lo unico que hay 
      que tener en cuenta es que en el archivo constant.py hay que dejat registrados esos nombres en las definiciones de las clases 
      HTMLElmnt y CSSClass
    - Eliminar todas las funciones de javascript que procesan eventos de mouse o keyboard dejando para los modulos de python esta tarea
    - En caso de agregar un modal, crear el/los elemento/s html (cuadro de dialogo, botones) y las clases para que aparezca/desaparezca, 
      informarlas en el constant.py y luego desde python se trabajará la lógica correspondiente
    - Ya saqué los binds de los botones de flechas, se pueden eliminar del html
    - Ya está hecha en python la logica de esconder la 'gameTitleScreen' luego de presionar una tecla o clickear con el mouse
    - Cuando el mouse tiene la forma de la mira (guided teleport) las coordenadas son las de el angulo superior izquierdo y no las de el
      centro de la mira. Si vos no podes corregir eso o decirme como obtener las coordenadas del centro, puedo corregir eso a mano pero 
      necesito saber las dimensiones de la mira.
    - Puse en el HTML los objetos para hacer los sonidos. Fijate si los puse en el lugar correcto y si los id siguen la convension de 
      nombres acostumbrada (en realidad lo mismo deberías hacer para todos los elementos ya que ahora les podes poner el id que quieras 
      mientras lo registres en constant.py)
    - OJO:
       - En modo guided-teleport, al pasar el mouse por algunas zonas del tablero, desaparece la mira y aparece la flecha comun o 
         incluso la barra vertical de ingreso de texto!! En esas condiciones al hacer click NO ejecuta la teletransportacion. Quiza 
         haya que asociar la clase 'separation-wrapper--guided-teleport-cursor' a algun otro elemento HTML. Eso lo podes probar vos 
         cambiando en las clases del constant.py tanto el elemento como la clase a utilizar para esta tarea
       - Hace un tiempo esta dando un warning como pagina insegura al cargarla desde internet, tal vez haya algun tipo de codigo 
         javascript que la haga "sospechosa" para el antivirus

### Notes

  - #### **Para trabajar tranquilo sin tener que cerrar cartel de CHASE_ siempre que se refresaca la pagina simplemente hay sacarlse la clase `game-title--active` al componente con clase `game-title`.** 
  - Para pintar los circulos (**`count-indicator__circle`**) hay que agregarles la clase **`count-indicator__circle--active`**. _Tienen un id para poder refernciarlos en Python_.
  - Sacar los botones del **`table`** para que no rompa.
  - Para terminar la sección de presentación del juego habría que agregar alguna variable que se inicialice en **`False`** y con la primera interación se cambie a **`True`**. La primera interacción la interpreté como `click` o `keydown`. Eso es para que mientras la sección de presentación del juego esté activa no se pueda interactuar con el juego. Entonces habría que preguntar si esa variable es **`True`** para las interacciónes con el juego. Eso es una idea.
  - Para implementar el cursor de guided teleport activo hay que agregarle la clase **`separation-wrapper--guided-teleport-cursor`** al componente con id **`separationWrapper`**.


#### 12/05/2022 - Nico
## Added
  - Mejoras en los iconos.
  - Tooltip para los botones de acciones.
  - Sección de instrucciones.
  - Sección de presentación. `Ver las notas para impletarlo`.
  - Ya saqué el **`svg_root`**.
  - Agregué el cursor para cuando guided teleport está activo. `Ver las notas para impletarlo`.


ui.py
import time
import os
import math
from constant import CSSClass, HTMLElmnt, Metrics, Tools, Anim, ToolHTMLElement

try:
    import browser
    from browser import document, svg, alert, timer, window, html
    # I'm running in a web browser => use Brython to manage the GUI (https://brython.info/)
    BROWSER = True
except:
    # Brython isn't available so use standard python text based interfase
    import os
    BROWSER = False

class UI:

    nullFunction = lambda *args, **kwargs: None

    # override all or any of these functions depending on the type of interfase, text (standard os console) or graphic (via Brython)
    resetAnim = nullFunction
    nextStepTime = nullFunction
    draw = nullFunction
    translate = nullFunction
    delete = nullFunction
    
    refreshScores = nullFunction
    refreshButtons = nullFunction
    refreshGuided = nullFunction
    refreshSafeness = nullFunction

    sndFire = nullFunction
    sndGetTool = nullFunction
    sndLevelUp = nullFunction
    sndLost = nullFunction

    askNewGame = nullFunction

    textDisplayBoard = nullFunction


class NoGui(UI):
    # when testing the game in text mode (no graphics available) use this class instead to manage the user interfase

    def askNewGame(self):
        again = ' '
        while again not in 'yns':
            again = input('Play again (y/n)? ').lower()
        return again in 'ys'

    def textDisplayBoard(self, board):

        os.system('cls')
        print()
        print('Foes left', board.foeCount, ' '*20, 'Score:', board.score, ' '*20, 'High Score:', board.highScore)
        print()
        print('  ' + ('. 1 2 3 4 5 6 7 8 9 ' * 20)[ : board.maxC*2 ] + '  ' )
        for row in range(board.maxR):
            print( (str(row%10) if row%10 != 0 else '.') + ' ', end='')
            for col in range(board.maxC):
                if board.grid[row][col] != None:
                    print(board.grid[row][col].char+' ', end='')
                else:
                    print('  ', end='')
            print( (str(row%10) if row%10 != 0 else '.') )
        print('  ' + ('. 1 2 3 4 5 6 7 8 9 ' * 20)[ : board.maxC*2 ] + '  ' )
        print()
        for t in board.toolStock:
            print( t + ' ' + str(board.toolStock[t] if board.toolStock[t] >= 0 else ''), end='   ' )
        print()
        print()


class GUI(UI):
    # class to manage all the Graphical User Interfase based on Brython and the browser objects

    SMALL_MOVE = 1
    BIG_MOVE = 15
    FAST_MOVE = 250  # milliseconds between clicks in a double click

    def __init__(self, board):

        self.board = board

        self.field = browser.document[HTMLElmnt.BATTLE_FIELD]

        self.animWhen = 0
        
        self.width =  self.field.offsetWidth
        self.height =  self.field.offsetHeight
        print(f'w:{self.width}, h:{self.height}')
        self.cellWidth = self.width / self.board.maxC
        self.cellHeight = self.height / self.board.maxR
        print('cell width, height:', self.cellWidth, self.cellHeight)
        self.relativeTop  = 0
        self.relativeLeft = 0
        self.absoluteTop  = self.field.abs_top  + self.relativeTop
        self.absoluteLeft = self.field.abs_left + self.relativeLeft

        print('rel top left', self.relativeTop, self.relativeLeft)
        print('abs top left', self.absoluteTop, self.absoluteLeft)
        
        self.mouseDownX = None
        self.mouseDownY = None 
        self.mouseUpX = None
        self.mouseUpY = None
        self.lastPointerMoveTimeStamp = 0

        self.keydownArrows = {'ArrowUp':False, 'ArrowDown':False, 'ArrowLeft':False, 'ArrowRight':False}
        self.playing = False

        self.bindsTitleScreen(bind=True)

    def bindsTitleScreen(self, bind):
        if bind:
            window.bind("mouseup", self.removeTitleAndStart)
            window.bind('touchend', self.removeTitleAndStart)
            window.bind('keyup', self.removeTitleAndStart)
        else:
            window.removeEventListener("mouseup", self.removeTitleAndStart)
            window.removeEventListener('touchend', self.removeTitleAndStart)
            window.removeEventListener('keyup', self.removeTitleAndStart)


    def removeTitleAndStart(self, evt):
        if browser.document[HTMLElmnt.TITLE_SCREEN_DIALOG].classList.contains(CSSClass.TITLE_SCREEN_ACTIVE):
            browser.document[HTMLElmnt.TITLE_SCREEN_DIALOG].classList.remove(CSSClass.TITLE_SCREEN_ACTIVE)
            self.bindsTitleScreen(bind=False)
            self.bindsGamePlay(bind=True)
            self.playing = True


    def bindsGamePlay(self, bind):
        if bind:
            self.field.bind("mousedown", self.pointerStart)
            self.field.bind("mousemove", self.pointerMove)
            self.field.bind("mouseup", self.pointerEnd)
            self.field.bind('touchstart', self.pointerStart)
            self.field.bind('touchmove', self.pointerMove)
            self.field.bind('touchend', self.pointerEnd)
            window.bind('keydown', self.keyPressed)
            window.bind('keyup', self.keyPressed)
            browser.document[HTMLElmnt.INSTRUCTIONS_TOGGLE].bind( 'click', lambda evt: self.showInstructions(True) )
            browser.document[HTMLElmnt.TELEPORT_BUTTON].bind( 'click', lambda evt: self.board.teleport() )
            browser.document[HTMLElmnt.SAFE_TELEPORT_BUTTON].bind( 'click', lambda evt: self.board.teleport(safe=True) )
            browser.document[HTMLElmnt.GUIDED_TELEPORT_BUTTON].bind( 'click', lambda evt: self.board.setGuided('toggle') )
            browser.document[HTMLElmnt.SMALL_BOMB_BUTTON].bind( 'click', lambda evt: self.board.bomb(big=False) )
            browser.document[HTMLElmnt.BIG_BOMB_BUTTON].bind( 'click', lambda evt: self.board.bomb(big=True) )
            browser.document[HTMLElmnt.NEW_GAME_BUTTON].bind( 'click', lambda evt: self.board.newGame() )
        else:
            self.field.removeEventListener("mousedown", self.pointerStart)
            self.field.removeEventListener("mousemove", self.pointerMove)
            self.field.removeEventListener("mouseup", self.pointerEnd)
            self.field.removeEventListener('touchstart', self.pointerStart)
            self.field.removeEventListener('touchmove', self.pointerMove)
            self.field.removeEventListener('touchend', self.pointerEnd)
            window.removeEventListener('keydown', self.keyPressed)
            window.removeEventListener('keyup', self.keyPressed)
            browser.document[HTMLElmnt.INSTRUCTIONS_TOGGLE].removeEventListener( 'click', lambda evt: self.showInstructions(True) )
            browser.document[HTMLElmnt.TELEPORT_BUTTON].removeEventListener( 'click', lambda evt: self.board.teleport() )
            browser.document[HTMLElmnt.SAFE_TELEPORT_BUTTON].removeEventListener( 'click', lambda evt: self.board.teleport(safe=True) )
            browser.document[HTMLElmnt.GUIDED_TELEPORT_BUTTON].removeEventListener( 'click', lambda evt: self.board.setGuided('toggle') )
            browser.document[HTMLElmnt.SMALL_BOMB_BUTTON].removeEventListener( 'click', lambda evt: self.board.bomb(big=False) )
            browser.document[HTMLElmnt.BIG_BOMB_BUTTON].removeEventListener( 'click', lambda evt: self.board.bomb(big=True) )
            browser.document[HTMLElmnt.NEW_GAME_BUTTON].removeEventListener( 'click', lambda evt: self.board.newGame() )

    def showInstructions(self, show=True):
        if show:
            self.playing = False
            self.bindsGamePlay(bind=False)
            browser.document[HTMLElmnt.INSTRUCTIONS].classList.remove(CSSClass.HIDE)
            browser.document[HTMLElmnt.INSTRUCTIONS_CLOSE].bind( 'click', lambda evt: self.showInstructions(False) )
        else:
            browser.document[HTMLElmnt.INSTRUCTIONS].classList.add(CSSClass.HIDE)
            browser.document[HTMLElmnt.INSTRUCTIONS_CLOSE].removeEventListener( 'click', lambda evt: self.showInstructions(False) )
            self.bindsGamePlay(bind=True)
            self.playing = True

    def rowcol2coords(self, row, col, relative=True):
        top  = self.relativeTop  if relative else self.absoluteTop
        left = self.relativeLeft if relative else self.absoluteLeft
        return ( int(left + col * self.cellWidth), int(top + row * self.cellHeight) )

    def coords2rowcol(self, x, y, relative=False):
        top  = self.relativeTop  if relative else self.absoluteTop
        left = self.relativeLeft if relative else self.absoluteLeft
        row = round((y - top - self.cellHeight/2) / self.cellHeight)
        col = round((x - left - self.cellWidth/2) / self.cellWidth)
        row = min( max(0,row), self.board.maxR - 1)
        col = min( max(0,col), self.board.maxC - 1)
        return ( row, col )

    def sndFire(self):
        timer.set_timeout(browser.document[HTMLElmnt.SND_FIRE].play, self.animWhen + Anim.STEP_TIME)

    def sndGetTool(self):
        timer.set_timeout(browser.document[HTMLElmnt.SND_GET_TOOL].play, self.animWhen)

    def sndLevelUp(self):
        timer.set_timeout(browser.document[HTMLElmnt.SND_LEVEL_UP].play, self.animWhen)

    def sndLost(self):
        timer.set_timeout(browser.document[HTMLElmnt.SND_LOST].play, self.animWhen + Anim.STEP_TIME)

    def resetAnim(self):
        self.animWhen = 0

    def nextStepTime(self):
        self.animWhen += Anim.STEP_TIME

    def _draw(self, img):
        print(f'_draw {img}')
        self.field <= img

    def _translate(self, id, x, y):
        print(f'_translate, id={id} en x:{x}, y:{y}')
        o = browser.document[id]
        o.style['top'] = y
        o.style['left'] = x

    def _delete(self, id):
        print(f'_delete, id={id}')
        browser.document[id].remove()


    def draw(self, boardObj):
        print(f'draw {boardObj.char}, id={boardObj.id} en {boardObj.row}, {boardObj.col}')
        x, y = self.rowcol2coords( boardObj.row, boardObj.col )
        print(f'    x:{x}, y:{y}')
        img = html.IMG(
            id = boardObj.id,
            src = boardObj.shape, 
            alt = boardObj.char,
            Class = 'board-object', 
            style = {
                'height': int(self.cellHeight),
                'width': int(self.cellWidth),
                'top': y, 
                'left': x,
            }
        )        
        timer.set_timeout(self._draw, self.animWhen + Anim.STEP_TIME, img)

    def translate(self, boardObj):
        print(f'translate, id={boardObj.id} en {boardObj.row}, {boardObj.col}')
        x, y = self.rowcol2coords( boardObj.row, boardObj.col )
        timer.set_timeout(self._translate, self.animWhen, boardObj.id, x, y)

    def delete(self, boardObj):
        print(f'delete, id={boardObj.id} en {boardObj.row}, {boardObj.col} type={type(boardObj)}')
        timer.set_timeout(self._delete, self.animWhen + Anim.STEP_TIME, boardObj.id)

    def pointerStart(self, evt):
        print(f'{evt.type} {evt.target.id}')
        if evt.type == 'mousedown':
            x, y = evt.x, evt.y
        elif evt.type == 'touchstart':
            evt.preventDefault()
            if len(evt.touches) == 1:
                x, y = evt.touches[0].pageX, evt.touches[0].pageY
            else:
                self.mouseDownX, self.mouseDownY = None, None  # cancel move
                return
        print(f'     x: {x}, y: {y}, coords: {self.coords2rowcol(x, y)}')
        self.mouseDownX, self.mouseDownY = x, y
        self.mouseUpX, self.mouseUpY = x, y  # if there is no movement the ponterMove will not be fired...
        
    def pointerMove(self, evt):
        print(f'{evt.type} {evt.target.id}')
        if evt.type == 'mousemove':
            x, y = evt.x, evt.y
        elif evt.type == 'touchmove':
            evt.preventDefault()
            if len(evt.touches) == 1:
                x, y = evt.touches[0].pageX, evt.touches[0].pageY
            else:
                self.mouseUpX, self.mouseUpY = None, None  # cancel move
                return
        print(f'     x: {x}, y: {y}, coords: {self.coords2rowcol(x, y)}')
        self.mouseUpX, self.mouseUpY = x, y

    def pointerEnd(self, evt):
        print(f'{evt.type} {evt.target.id}')
        print(f'    {self.mouseDownX}, {self.mouseDownY} -> {self.mouseUpX}, {self.mouseUpY}')
        print(f'last timestamp: {self.lastPointerMoveTimeStamp}, current: {evt.timeStamp}, dif: {evt.timeStamp - self.lastPointerMoveTimeStamp}')
        if evt.type == 'touchmove':
            evt.preventDefault()

        deltaX = self.mouseUpX - self.mouseDownX
        deltaY = self.mouseUpY - self.mouseDownY
        length = math.sqrt( deltaX**2 + deltaY**2 )
        print(f'deltaX: {deltaX}, deltaY: {deltaY}, length: {length}')
        if (length/self.cellWidth) < GUI.SMALL_MOVE:
            deltaC, deltaR = 0, 0
        else:
            angle = math.asin(deltaY/length) if deltaX > 0 else math.pi - math.asin(deltaY/length)
            angle = angle + 2*math.pi if angle < 0 else angle
            cuadrant = ( round((angle / math.pi) * 4) / 4 ) * math.pi
            print(f'Angle: {angle/math.pi}*pi, cuadrant: {cuadrant/math.pi}*pi')
            print(f'sin(cuadrant): {math.sin(cuadrant)}, cos(cuadrant): {math.cos(cuadrant)}')
            sign = lambda x: 0 if x == 0 else ( 1 if x > 0 else -1 )
            deltaR = sign(round(math.sin(cuadrant),3))
            deltaC = sign(round(math.cos(cuadrant),3))

        if self.board.guided:
            if deltaR == 0 and deltaC == 0:
                self.board.teleport(guided=True, coords=self.coords2rowcol(self.mouseUpX, self.mouseUpY))
        else:
            repeat = evt.shiftKey or \
                     (evt.timeStamp - self.lastPointerMoveTimeStamp < GUI.FAST_MOVE) or \
                     ((length/self.cellWidth) > GUI.BIG_MOVE)
            self.board.move(deltaR,deltaC, repeat)
            self.lastPointerMoveTimeStamp = evt.timeStamp

    def keyPressed(self, evt):
        if not self.playing:
            return
        if evt.type == 'keydown':
            print('keydown', evt.key)
            if evt.key in self.keydownArrows:
                self.keydownArrows[evt.key] = True
            return

        if evt.type == 'keyup':
            print('keyup', evt.key)
            if evt.key in '123456789 ' or evt.key in ['Home', 'End', 'PageUp', 'PageDown', 'Clear']:
                deltaR, deltaC = 0, 0
                if evt.key in '741' or evt.key in ['Home', 'End']:
                    deltaC = -1
                elif evt.key in '963' or evt.key in ['PageUp', 'PageDown']:
                    deltaC = 1
                if evt.key in '789' or evt.key in ['Home', 'PageUp']:
                    deltaR = -1
                elif evt.key in '123' or evt.key in ['End', 'PageDown']:
                    deltaR = 1
                self.board.move(deltaR, deltaC, evt.shiftKey)
            elif evt.key in self.keydownArrows:
                if not self.keydownArrows[evt.key]:
                    return # this is the keyup of an arrow that has been processed in combination with another
                deltaR, deltaC = 0, 0
                if self.keydownArrows['ArrowUp'   ]:
                    deltaR = deltaR - 1 
                if self.keydownArrows['ArrowDown' ]:
                    deltaR = deltaR + 1 
                if self.keydownArrows['ArrowLeft' ]:
                    deltaC = deltaC - 1 
                if self.keydownArrows['ArrowRight']:
                    deltaC = deltaC + 1 
                self.keydownArrows = {k: False for k in self.keydownArrows}
                print('move', deltaR, deltaC)
                self.board.move(deltaR, deltaC, evt.shiftKey)
            elif evt.key in 'tT':
                self.board.teleport()
            elif evt.key in 'sS':
                self.board.teleport(safe=True)
            elif evt.key in 'gG':
                self.board.setGuided('toggle')
            elif evt.key in 'vV':
                self.board.bomb(big=False)
            elif evt.key in 'bB':
                self.board.bomb(big=True)
            elif evt.key in 'nN':
                self.board.newGame()
            return


    def refreshScores(self):
        if self.board.puzzle:
            browser.document[HTMLElmnt.DIV_LEVEL].classList.add(CSSClass.HIDE)
            browser.document[HTMLElmnt.DIV_FOE_COUNT].classList.remove(CSSClass.HIDE)
            browser.document[HTMLElmnt.DIV_SCORE].classList.add(CSSClass.HIDE)
            browser.document[HTMLElmnt.DIV_STEPS].classList.remove(CSSClass.HIDE)
            browser.document[HTMLElmnt.DIV_HIGH_SCORE].classList.add(CSSClass.HIDE)
        else:
            browser.document[HTMLElmnt.DIV_LEVEL].classList.remove(CSSClass.HIDE)
            browser.document[HTMLElmnt.DIV_FOE_COUNT].classList.remove(CSSClass.HIDE)
            browser.document[HTMLElmnt.DIV_SCORE].classList.remove(CSSClass.HIDE)
            browser.document[HTMLElmnt.DIV_STEPS].classList.remove(CSSClass.HIDE)
            browser.document[HTMLElmnt.DIV_HIGH_SCORE].classList.remove(CSSClass.HIDE)

        browser.document[HTMLElmnt.TEXT_LEVEL].textContent = self.board.level
        browser.document[HTMLElmnt.TEXT_FOE_COUNT].textContent = self.board.foeCount
        browser.document[HTMLElmnt.TEXT_SCORE].textContent = self.board.score
        browser.document[HTMLElmnt.TEXT_STEPS].textContent = self.board.steps
        browser.document[HTMLElmnt.TEXT_HIGH_SCORE].textContent = self.board.highScore

    def refreshButtons(self, tool=None):
        tools = self.board.toolStock if tool is None else {tool}
        for t in tools:
            print('refresh buttons', t, self.board.toolStock[t])
            if self.board.toolStock[t] < 0:
                browser.document[ToolHTMLElement[t]].classList.add(CSSClass.HIDE)
            elif self.board.toolStock[t] == Metrics.TOOL_INFINITE:
                browser.document[ToolHTMLElement[t]].classList.remove(CSSClass.HIDE)
                browser.document[HTMLElmnt.INFINITE_ID_PREFIX + ToolHTMLElement[t]].classList.remove(CSSClass.HIDE)
                browser.document[HTMLElmnt.CIRCLES_PARENT_ID_PREFIX + ToolHTMLElement[t]].classList.add(CSSClass.HIDE)
            else:
                browser.document[ToolHTMLElement[t]].classList.remove(CSSClass.HIDE)
                browser.document[HTMLElmnt.INFINITE_ID_PREFIX + ToolHTMLElement[t]].classList.add(CSSClass.HIDE)
                browser.document[HTMLElmnt.CIRCLES_PARENT_ID_PREFIX + ToolHTMLElement[t]].classList.remove(CSSClass.HIDE)
                for c in range(Metrics.MAX_TOOL_STOCK):
                    idCircle = ToolHTMLElement[t] + HTMLElmnt.CIRCLES_ID_SUFFIX + str(c+1)
                    if self.board.toolStock[t] > c:
                        browser.document[idCircle].classList.add(CSSClass.COUNT_INDICATOR)
                    else:
                        browser.document[idCircle].classList.remove(CSSClass.COUNT_INDICATOR)

    def refreshGuided(self, value):
        if value:
            browser.document[HTMLElmnt.GUIDED_TELEPORT_CURSOR_SCOPE].classList.add(CSSClass.GUIDED_TELEPORT_CURSOR)
            browser.document[HTMLElmnt.GUIDED_TELEPORT_BUTTON].classList.add(CSSClass.SELECTED_BUTTON)
        else:
            browser.document[HTMLElmnt.GUIDED_TELEPORT_CURSOR_SCOPE].classList.remove(CSSClass.GUIDED_TELEPORT_CURSOR)
            browser.document[HTMLElmnt.GUIDED_TELEPORT_BUTTON].classList.remove(CSSClass.SELECTED_BUTTON)

    def refreshSafeness(self):
        browser.document[HTMLElmnt.TEXT_SAFENESS].textContent = "{:.4f}".format(self.board.safeness)

    def askNewGame(self):
        return browser.confirm('Play again?')



js/app.js
const fieldWrapper = document.querySelector('.field-wrapper');
const field = document.getElementById('field');
const instructions = document.getElementById('instructions');
const pressToEnter = document.getElementById('pressToEnter');
const gameTitleScreen = document.getElementById('gameTitleScreen');
const newGameButton = document.getElementById('new');
const fieldWidth = field.offsetWidth;

loadDimensions = () => {
  window.innerWidth < window.innerHeight ? fieldWrapper.style.height = claculateDimension(22, 26) : fieldWrapper.style.height = claculateDimension(26, 22);
  console.log('fieldWidth', fieldWidth);
  console.log(fieldWrapper.offsetHeight);
  console.log('Relación: ', fieldWidth * 22 / 26);
}

claculateDimension = (d1, d2) => {
  return `${(fieldWidth / d1) * d2}px`;
}

retainScroll = () => {
  window.addEventListener("keydown", e => {
    if(["Space","ArrowUp","ArrowDown","ArrowLeft","ArrowRight"].indexOf(e.code) > -1) {
        e.preventDefault();
    }
  }, false);
}

// console.log('------------------------------------------------>');
loadDimensions();
retainScroll();
// console.log('------------------------------------------------>');


constant.py
class HTMLElmnt:
    BATTLE_FIELD                  = 'field'
    SMALL_BOMB_BUTTON             = 'v'
    BIG_BOMB_BUTTON               = 'b'
    TELEPORT_BUTTON               = 't'
    SAFE_TELEPORT_BUTTON          = 's'
    GUIDED_TELEPORT_BUTTON        = 'g'
    CIRCLES_ID_SUFFIX             = 'Circle'
    CIRCLES_PARENT_ID_PREFIX      = 'count-indicator-'
    INFINITE_ID_PREFIX            = 'count-infinite-'
    NEW_GAME_BUTTON               = 'new'
    GUIDED_TELEPORT_CURSOR_SCOPE  = 'separationWrapper'
    TITLE_SCREEN_DIALOG           = 'gameTitleScreen'
    TEXT_LEVEL                    = 'level'
    DIV_LEVEL                     = 'div-level'
    TEXT_FOE_COUNT                = 'foeCount'
    DIV_FOE_COUNT                 = 'div-foeCount'
    TEXT_SCORE                    = 'score'
    DIV_SCORE                     = 'div-score'
    TEXT_HIGH_SCORE               = 'highScore'
    DIV_HIGH_SCORE                = 'div-highScore'
    TEXT_SAFENESS                 = 'safeness'
    DIV_SAFENESS                  = 'div-safeness'
    TEXT_STEPS                    = 'steps'
    DIV_STEPS                     = 'div-steps'
    SND_FIRE                      = 'sndFire'
    SND_GET_TOOL                  = 'sndGetTool'
    SND_LEVEL_UP                  = 'sndLevelUp'
    SND_LOST                      = 'sndLost'
    INSTRUCTIONS_TOGGLE           = 'instructions-toggle'
    INSTRUCTIONS                  = 'instructions'
    INSTRUCTIONS_CLOSE            = 'instructions-close'



class CSSClass:
    BOARD_OBJECT                  = 'board-object'
    COUNT_INDICATOR               = 'count-indicator__circle--active'
    GUIDED_TELEPORT_CURSOR        = 'separation-wrapper--guided-teleport-cursor'
    SELECTED_BUTTON               = 'button-selected'
    HIDE                          = 'hide'
    TITLE_SCREEN_ACTIVE           = 'game-title--active'
    INSTRUCTIONS_ACTIVE           = 'instructions--active'


class Tools:
    # Types of tools that the Hero can find in the Board.
    SMALL_BOMB      = 'SmalBomb'
    BIG_BOMB        = 'BigBomb'
    TELEPORT        = 'Teleport'
    SAFE_TELEPORT   = 'SafeTeleport'
    GUIDED_TELEPORT = 'GuidedTeleport'

ToolHTMLElement = {
    Tools.SMALL_BOMB:      HTMLElmnt.SMALL_BOMB_BUTTON,
    Tools.BIG_BOMB:        HTMLElmnt.BIG_BOMB_BUTTON,
    Tools.TELEPORT:        HTMLElmnt.TELEPORT_BUTTON,
    Tools.SAFE_TELEPORT:   HTMLElmnt.SAFE_TELEPORT_BUTTON,
    Tools.GUIDED_TELEPORT: HTMLElmnt.GUIDED_TELEPORT_BUTTON,
}

class Prio:
    # Priority to remain in a cell of the grid when two BoardObject run into each other.
    LOWER     = 0
    HERO      = 10
    FOE       = 50
    FIRE      = 80
    DEAD_HERO = 90

class TextChar:
    HERO            = '\u00B7'
    DEAD_HERO       = 'X'
    FOE             = '\u03A9'
    FIRE            = '\u039E'
    SMALL_BOMB      = 'v'
    BIG_BOMB        = 'b'
    SAFE_TELEPORT   = 's'
    GUIDED_TELEPORT = '\u0398'

class GraphShape:
    HERO            = 'assets/img/user-green.svg'
    DEAD_HERO       = 'assets/img/user-red.svg'
    FOE             = 'assets/img/omega.svg'
    FIRE            = 'assets/img/fire.svg'
    SMALL_BOMB      = 'assets/img/small-bomb.svg'
    BIG_BOMB        = 'assets/img/big-bomb.svg'
    SAFE_TELEPORT   = 'assets/img/safe-teleport.svg'
    GUIDED_TELEPORT = 'assets/img/guided-teleport.svg'

class Metrics:
    # metrics that will determine how difficult the game will be to play
    BOARD_DIM = (26,22)
    INIT_FOE_COUNT = 0
    INCREMENT_FOE_COUNT_BY_LEVEL = 10
    DROP_TOOL_MU = 0
    DROP_TOOL_SIGMA = 1
    MAX_TOOL_STOCK = 5
    TOOL_INFINITE = 100
    TOOL_DISABLE = -1
    INIC_TOOL_STOCK = {
        Tools.TELEPORT: TOOL_INFINITE,
        Tools.SAFE_TELEPORT: 3,
        Tools.GUIDED_TELEPORT: 3,
        Tools.SMALL_BOMB: 3,
        Tools.BIG_BOMB: 3,            
    }
    DIE_BEYOND_EDGES = False

class Anim:
    # times involved in animation (in ms)
    STEP_TIME = 50
